### PR TITLE
Spec file structure

### PR DESCRIPTION
**Pull request by Kegsay**
This PR modifies the file naming format for RST files in `/specification` to be more sane and track the sections/subsections they are supposed to reside in. It also modifies `gendoc.py` to enforce this naming convention in order to reduce the prevalence of silent RST errors, sections appearing in unintended places, etc. The new filenames in `/specification` are:
```
00_00_intro.rst
01_00_client_server_api.rst
01_01_typing_notifications.rst
01_02_receipts.rst
01_03_history_visibility.rst
02_00_events.rst
02_01_voip_events.rst
02_02_event_signing.rst
03_00_application_service_api.rst
04_00_server_server_api.rst
05_00_content_repository.rst
06_00_end_to_end_encryption.rst
07_00_push_overview.rst
07_01_push_cs_api.rst
07_02_push_push_gw_api.rst
08_00_address_book_repo.rst
09_00_identity_servers.rst
10_00_appendices.rst

```

### CODE REVIEW

**Comment by illicitonion** [comment]
LGTM

File: specification/01_01_typing_notifications.rst [hunk 1]
```diff
@@ -39,7 +39,7 @@ users who are not currently typing, as that list gets big quickly. The client
 should mark as not typing, any user ID who is not in that list.
 
 Server APIs
```
Comment (ara4n) [inline specification/01_01_typing_notifications.rst]: Isn't it going to be weird to have s2s-specific stuff poping up in the "01 client server API" section?

**Comment by ara4n** [comment]
Other than querying why we're moving some of the "grouped by function" stuff like receipts & typing notifs into the CS API section, LGTM.  What's the problem with having these slightly extensiony things separated out from the 'core' CS API and grouped together by function?  I hate the idea of scattering them over the spec into separate CS/SS/Event sections, but I'm also not keen on inserting them into the CS API section.. (which was why they were in their own appendixy chapters anyway).  I wonder if we should define them as extensions of some flavour anyway, albeit ones mandated in a baseline interoperability profile.

**Comment by Kegsay** [comment]
The problem with "grouped by function" is it serves no-one well. The section ends up trying to serve two different audiences and forces each audience to read the *entire* spec (rather than the section they are interested in, e.g. Client-Server or Server-Server) in order to grok it.

I've spoken with @illicitonion at length about how best to approach this, and we're converging on a solution along the lines of:
```
0_overview.rst            // Contains object model, 'Events' section
1_client_server_v1.rst    // HTTP APIs, explanation of pagination
2_client_server_v2.rst
3_federation.rst
4_app_services.rst
5_push.rst
6_content_repo.rst
7_identity_server.rst
8_appendices.rst
```

This partitions the spec cleanly down the audience's goals. Yes, this does mean to some extent there will be some elements which appear in both CS/SS/Events, but we can always heavily use internal links to link sections together. Ultimately, we can't represent the spec in a way to please everyone (unless we template *all the things* and have a 'Spec by function' or 'Spec by protocol').

**Comment by ara4n** [comment]
More thoughts:

1. This refactor breaks a whole lot of URLs which are out there in the wild, e.g. in https://github.com/matrix-org/sygnal.  We should change them to link to anchors in the main spec to protect them from further refactoring.

2. I'm not convinced at the proposed new structure in the above comment, as:
 * if we are incrementally migrating from v1 to v2, i'm not sure it makes sense to keep them as separate monolithic chunks of spec.  It'd be better to just move obsolete v1 stuff into an obsolete appendix somewhere and add new APIs e.g. v2 into the same overall 'client server' section.
 * App services as more important to most users than Federation, as we expect more folks to want to write and understand ASes than HSes.  So I'd put it first.

3. I think we're also somehow talking cross purposes on the meaning of "grouped by function".  I was concerned that in this PR we take a bunch of related APIs for a function (i.e. both CS and SS considerations for typing notifs) and wedge them all into the CS section of the spec, which feels like it's mis-filed.  However, splitting them up between CS & SS sections is also bad, as it scatters the useful info.

My suggestion (after chatting a bit with Mark about it at the end of Friday) is a bit more radical: perhaps we should just formally modularise these features into well-defined modules within the spec (note deliberately avoiding the word extension :).  In other words, we'd probably have a section between the CS and AS chapters called "Modules" or something, which contains a bunch of chapters, one per non-core feature (typing notifs, delivery reports, presence, etc). Additionally, we add a chapter to the Overview section called "Feature Profiles" which defines precisely which modules are mandatory for which client and server use cases - it's a complete screwup that the spec doesn't make this explicit currently (https://matrix.org/jira/browse/SPEC-219).

So, the "Feature Profiles" section would be a table showing something like:
 * "GUI desktop client profile": all modules, with the exception of: optional push notifs
 * "Mobile client profile": all modules, with push notifs mandatory
 * "CLI desktop client profile": everything but VoIP and Video
 * "Standalone web client": everything
 * "Embedded web client": all the non-core modules are optional
 * "IOT device": all the non-core modules are optional

The actual non-core modules that could be split out into the Modules section would be something like:
 * Presence
 * Typing notifs
 * Read receipts
 * E2E
 * Content repo
 * VoIP
 * Conferencing 

This leaves the core as the basic ideas of managing rooms, sending messages, handling event stream, etc - making the core much more manageable and digestible, and carefully encapsulating the various non-core modules and grouping their stuff into one well-defined place so they aren't scattered around.

Meanwhile, this also gives us a nice way of getting people to draft and even contribute 3rd party modules - i.e. "format it to look like one of our modules".

Before anyone panics and thinks i've reinvented XEPs, the huge difference is that by making Feature Profiles a first class citizen, we're hopefully controlling and limiting fragmentation from the outset.  It's explicitly a bug to write a desktop client that doesn't implement the specified set of module functionality.  And meanwhile the baseline set of modules we provide are all nicely engineered coherently to play well together from the outset.

WDYT?

**Comment by Kegsay** [comment]
I... like it. Splitting the spec into sections relevant to the audience (desktop, mobile, iot, etc) makes me very happy. My immediate concern was that it was "reinvented xeps", but if we are explicitly enforcing that targets (desktop, mobile, etc) implement the modules we say, this could work *extremely well*. 

Great idea.

**Comment by Kegsay** [comment]
So these modules would contain the protocol speak for both CS and SS? This does somewhat gear the spec towards a client audience (pick what kind of client you're doing and do it) whereas server devs need to digest everything (which I'm not opposing, just pointing out). I can put a reworked draft layout on Monday and let it stew for you to look at (pending @illicitonion feedback).

**Comment by ara4n** [comment]
Woot.  Just to confirm: i'm not suggesting we split the spec into sections per-profile, but instead just have core + modules, and assign profiles to particular combinations of modules.

Each module would indeed have both details for CS and SS.  However, (and I forgot to mention this earlier; it was @NegativeMjark's idea), the tentative suggestion is that both the overall CS and SS sections and the per-module CS and SS sections could be re-characterised as "Stuff that client implementors need to know" and "Stuff server implementors need to know".  In practice, the contents stays largely the same: the CS API specifics would all be in the "Client Behaviour" section, and the SS API specifics would all be in the "Server Behaviour" section.  However, the tone of each section would be more targetted to the respective audiences... and there would inevitably be some overlap.  E.g. the "Server Behaviour" section would include a bunch of stuff about how servers should implement the relevant bits of the CS API, as well as all the SS magic.  Meanwhile, "Overview" would continue to have the high-level overview of the whole architecture - i.e. stuff that you need to know if you're writing clients *or* servers.

And yup, I'd love to review another layout - this is our chance to really get it right.

**Pull request merged**